review/get_review_ids.py

```diff
@@ -0,0 +1,212 @@
+
+    if len(problem_data) < num_desired:
+        return []
+
+    problems_to_show = random.sample(problem_data, num_desired)
+    review_course_id = REVIEW_COURSE_MAPPING[str(current_course)]
+    urls = []
+    for problem, correctness, attempts in problems_to_show:
+        urls.append((TEMPLATE_URL.format(course_id=review_course_id,
+                    type='problem', xblock_id=problem), correctness, attempts))
+    return urls
+
+def get_vertical(current_course):
+    '''
+    Looks through all the problems a learner has previously loaded and
+    finds their parent vertical. Then randomly selects a single vertical
+    to show the learner.
+
+    Parameters:
+        current_course (CourseLocator): The course the learner is currently in
+
+    Returns a url (str) with the vertical id to render for review.
+    '''
+    user = crum.get_current_user()
+
+    enroll_user(user, current_course)
+
+    store = modulestore()
+    course_usage_key = store.make_course_usage_key(current_course)
+    course_blocks = get_course_blocks(user, course_usage_key)
+
+    vertical_data = set()
+
+    # Each record corresponds to a problem the user has loaded
+    # in the original course
+    for record in StudentModule.objects.filter(**{'student_id': user.id, 'course_id': current_course, 'module_type': 'problem'}):
+        # record.module_state_key takes the form:
+        # block-v1:{course_id}+type@problem+block@{problem_id}
+        problem_id = str(record.module_state_key).split("@")[-1]
+
+        # Actual logic regarding the record should go here
+        state = json.loads(record.state)
+
+        # The key 'selected' shows up if a problem comes from a
+        # library content module. These cause issues so we skip this.
+        # Issue: Library content contains problems but the CSM brings up
+        # the library content and not the problems within
+        if 'selected' in state:
+            continue
+
+        # parent takes the form: block-v1:{course_id}+type@vertical+block@{vertical_id}
+        parent = course_blocks.get_parents(record.module_state_key)[0]
+        vertical_id = str(parent).split("@")[-1]
+
+        vertical_data.add(vertical_id)
+    if len(vertical_data) < 1:
+        return []
+
+    vertical_to_show = random.sample(vertical_data, 1)[0]
+    review_course_id = REVIEW_COURSE_MAPPING[str(current_course)]
+    return (TEMPLATE_URL.format(course_id=review_course_id,
+                    type='vertical', xblock_id=vertical_to_show))
+
+def enroll_user(user, current_course):
+    '''
+    If the user is not enrolled in the review version of the course,
+    they are unable to see any of the problems. This ensures they
+    are enrolled so they can see review problems.
+
+    Parameters:
+        user (User): the current user interacting with the review xBlock
+        current_course (CourseLocator): The course the learner is currently in
+    '''
+    enrollment_course_id = ENROLLMENT_COURSE_MAPPING[str(current_course)]
```

Comment:
So this XBlock works by having two courses? One from which the previously answered problems are pulled *and* one into which the review answers are recorded?

Reply:
Sort of. It looks at the problems a learner has originally interacted with in their actual course. Then it pulls a random sample of copies of those problems from the "review" course where the learner will answer those as part of this xBlock. This allows for those problems to have a clean slate from their original answer and be marked as ungraded with unlimited attempts.

Reply:
Question: What if the review XBlock (randomly) pulls problems that have already been offered and answered in the "review" course? Will the problems show up already answered with the student's previous answer?

Reply:
For right now, yes. I am actively working on that, but decided it was not a big enough problem to stop the initial release. Good eye. This has to do with that function that was blank and you told me to rip out for now.